src/core/utils/flash.cpp

```diff
@@ -0,0 +1,268 @@
+
+    if ((mSwapSize - record.GetSize()) < mSwapUsed)
+    {
+        Swap();
+        VerifyOrExit((mSwapSize - record.GetSize()) >= mSwapUsed, error = OT_ERROR_NO_BUFS);
+    }
+
+    otPlatFlashWrite(&GetInstance(), mSwapIndex, mSwapUsed, &record, record.GetSize());
+
+    record.SetAddCompleteFlag();
+    otPlatFlashWrite(&GetInstance(), mSwapIndex, mSwapUsed, &record, sizeof(RecordHeader));
+
+    mSwapUsed += record.GetSize();
+
+exit:
+    return error;
+}
+
+bool Flash::DoesValidRecordExist(uint32_t aOffset, uint16_t aKey) const
+{
+    RecordHeader record;
+    bool         rval = false;
+
+    for (; aOffset < mSwapUsed; aOffset += record.GetSize())
+    {
+        otPlatFlashRead(&GetInstance(), mSwapIndex, aOffset, &record, sizeof(record));
+
+        if (record.IsValid() && record.IsFirst() && (record.GetKey() == aKey))
+        {
+            ExitNow(rval = true);
+        }
+    }
+
+exit:
+    return rval;
+}
+
+void Flash::Swap(void)
+{
+    uint8_t  dstIndex  = !mSwapIndex;
+    uint32_t dstOffset = kSwapMarkerSize;
+    Record   record;
+
+    otPlatFlashErase(&GetInstance(), dstIndex);
+
+    for (uint32_t srcOffset = kSwapMarkerSize; srcOffset < mSwapUsed; srcOffset += record.GetSize())
+    {
+        otPlatFlashRead(&GetInstance(), mSwapIndex, srcOffset, &record, sizeof(RecordHeader));
+
+        VerifyOrExit(record.IsAddBeginSet());
+
+        if (!record.IsValid() || DoesValidRecordExist(srcOffset + record.GetSize(), record.GetKey()))
+        {
+            continue;
+        }
+
+        otPlatFlashRead(&GetInstance(), mSwapIndex, srcOffset, &record, record.GetSize());
+        otPlatFlashWrite(&GetInstance(), dstIndex, dstOffset, &record, record.GetSize());
+        dstOffset += record.GetSize();
+    }
+
+exit:
+    otPlatFlashWrite(&GetInstance(), dstIndex, 0, &sSwapActive, sizeof(sSwapActive));
+    otPlatFlashWrite(&GetInstance(), mSwapIndex, 0, &sSwapInactive, sizeof(sSwapInactive));
+
+    mSwapIndex = dstIndex;
+    mSwapUsed  = dstOffset;
+}
+
+otError Flash::Delete(uint16_t aKey, int aIndex)
+{
+    otError      error = OT_ERROR_NOT_FOUND;
+    int          index = 0;
+    RecordHeader record;
+
+    for (uint32_t offset = kSwapMarkerSize; offset < mSwapUsed; offset += record.GetSize())
+    {
+        otPlatFlashRead(&GetInstance(), mSwapIndex, offset, &record, sizeof(record));
+
+        if ((record.GetKey() != aKey) || !record.IsValid())
+        {
+            continue;
+        }
+
+        if (record.IsFirst())
+        {
+            index = 0;
+        }
+
+        if ((aIndex == index) || (aIndex == -1))
+        {
+            record.SetDeleted();
+            otPlatFlashWrite(&GetInstance(), mSwapIndex, offset, &record, sizeof(record));
+            error = OT_ERROR_NONE;
+        }
+
+        if ((index == 1) && (aIndex == 0))
+        {
+            record.SetFirst();
+            otPlatFlashWrite(&GetInstance(), mSwapIndex, offset, &record, sizeof(record));
+            ExitNow();
+        }
```

Comment:
Reading this logic I think in some (rare) corner cases, we may get to bad state. Here is situation I can think of:
- Suppose user has already called `Add` with a bunch of values for a key. Say there are 3 records: `r0(first), r1, r2` (first one is marked as `IsFirst`).
- User now calls `Set` for the same key, it would add a new record at the end with `IsFirst` set -> `r0(first), r1, r2, n0(first)`.
- Now user issues a `Delete` for index 0 (we can have a vulnerability here):
  - We run through the algo: `r0` is marked as deleted, `r1` is marked as first. We get to `n0` and it is marked as deleted.
  - Now the records in the flash are `r1(fist), r2.`
  - So we get some old entries back in the list.

The only way I can think of for us to get into such situation is if user mixes  `Set` and `Add` for the same key. I think we currently don't use this from core, but would be good to be aware of this limitation.

Reply:
Yes, thanks for catching that! I updated the implementation and unit tests to properly handle/check this scenario.

Reply:
How about if we later (future PR?) change the requirement for `Set` and `Add` (in `otPlatSettings{name}` as well) such we don't allow user to mix and match `Set` and `Add` (considering it as incorrect use and undefined behavior (not supported)). 

Basically expect a key to either be used as a list (use `Add`) or as a single value (use `Set`).

Thoughts?

Reply:
We can consider this in a future PR. Just curious, how would you enforce this?

Reply:
Not sure if there'd be an easy way to enforce it in the code. 
It would be just changing the documentation (and making sure we follow the new policy in core - which we already do).